src/minio/core/policy_builder.py

```
@@ -0,0 +1,189 @@
+            )
+
+        bucket_in_path, relative_path = path_parts
+
+        # Validate bucket matches configuration
+        if bucket_in_path != self.bucket_name:
+            raise PolicyOperationError(
+                f"Path bucket '{bucket_in_path}' does not match configured bucket '{self.bucket_name}'"
+            )
+
+        return relative_path.rstrip("/")
+
+    def _has_path_access(self, clean_path: str) -> bool:
+        """Check if the policy already grants access to the specified path."""
+        target_resource = f"arn:aws:s3:::{self.bucket_name}/{clean_path}/*"
+
+        for statement in self.policy_model.policy_document.statement:
+            if statement.effect == PolicyEffect.ALLOW:
+                resources = (
+                    statement.resource
+                    if isinstance(statement.resource, list)
+                    else [statement.resource]
+                )
+                if target_resource in resources:
+                    return True
+        return False
+
+    def _find_list_bucket_statement(self) -> PolicyStatement | None:
+        """Find the ListBucket statement with prefix conditions."""
+        for stmt in self.policy_model.policy_document.statement:
+            if (
+                PolicyAction.LIST_BUCKET
+                in (stmt.action if isinstance(stmt.action, list) else [stmt.action])
+                and stmt.condition
+                and "StringLike" in stmt.condition
+                and "s3:prefix" in stmt.condition["StringLike"]
+            ):
+                return stmt
+        return None
+
+    def _add_to_list_bucket_statement(self, clean_path: str) -> None:
+        """Add path prefixes to the ListBucket statement."""
+        list_bucket_stmt = self._find_list_bucket_statement()
+        if not list_bucket_stmt or not list_bucket_stmt.condition:
+            logger.warning("No ListBucket statement found to add path to")
+            return
```

Hm, should it not add a ListBucket statement in this case? This means the user will have access to the path but can't list the bucket to see that they have access to the path IIUC

---

our assumption is we only have one cdm-lake bucket - this means the bucket is not cdm-lake.

---

So should this throw an error then? IIUC it's adding a path to a policy that is pointing to the wrong bucket in the list buckets part of the policy

---

yea - I updated it to raise error instead of return. The bucket list statement should always be there as part of default policy creation. If it's missing, we should raise.